### DISCUSSION TITLE
Automatic terminfo propagation over SSH

### DISCUSSION THREAD

**Discussion by dgl** [ghostty-org/ghostty#4156] [Feature Requests, Ideas] [answered]
Needing to copy terminfo to remote systems is a rather FAQ. It would be possible for the shell integration to automatically inject an SSH config which copies the config for the user.

The following is a PoC, it obviously has an awfully hacky shell function for parsing the ssh arguments and the tracking of whether it has copied the terminfo over is the silliest thing. But it proves it can work, even with more complex ssh commands like `ssh -J jump some-host`.

A real implementation that does this should prompt the user before doing this on a remote system (maybe configurable via the shell integration options and also copy over the shell integration if configured to do so). It also deliberately wraps ssh as a shell function only, so this doesn't affect scripts.

In ~/.ssh/ghosty-config:

```
# Order matters, more specific has to be first
Match Exec "mkdir -p ~/.ssh-ghosty; ! [ -e $HOME/.ssh-ghosty/%h ]" Host *
  PermitLocalCommand yes
  LocalCommand ssh-ghosty-terminfo %h %r %p %j
  ControlMaster auto
  ControlPath ~/.ssh-ghosty/controlmaster-%h

Include ~/.ssh/config
# Using -F means it doesn't pick up the system config.
Include /etc/ssh/ssh_config
```

In ~/bin/ssh-ghosty-terminfo this messy script (and chmod +x it):

```
#!/bin/sh
host="$1"
user="$2"
port="$3"
jump="$4"
if [ "$user" = "root" ]; then
  # could be a prompt?
  echo Not installing TERMINFO as root, would be system-wide
  exit 0
fi
touch ~/.ssh-ghosty/$host
sleep 2
j=""
if [ "$jump" != "" ]; then
  j="-J $jump"
fi
infocmp -x | ssh -l "$user" -p $port $j $host -- tic -x -
```

Then in .zshrc / bashrc (this bit could be configured by the shell integration similar to how it does for sudo):

```
ssh() { if [[ "${*/ -F}" == "$*" ]]; then command ssh -F ~/.ssh/ghosty-config "$@"; else command ssh "$@"; fi }
```

**Comment by dacort** [comment]
I just got caught by this as well - I often SSH into random hosts (my `~/.ssh/known_hosts` has 167 entries, tho some are dupes of course for different key types) and copying over terminfo every time isn't reasonable. 

For now, I've just set `TERM=xterm-256color` in my `.ssh/config` since remote terminals are fairly broken otherwise.

**Reply by socheatsok78** [reply]
I guess you only need to install it system-wide then you don't need to worry to do it every time?

**Reply by dacort** [reply]
I should clarify that I mean "copying over terminfo on every new host I connect to" instead of "every time". I work with a lot of ephemeral EC2 instances so would need to copy over the terminfo to every new instance (or modify my company's provisioning scripts to deploy the terminfo, which won't happen), right?

**Comment by 00-kat** [comment, marked as answer]
Superseded by the (more broad in scope) #5892.
